pnpm install broke after 10.6.1

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

10.6.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

This is my docker file

FROM node:22-slim
ENV NODE_ENV production
ENV PUPPETEER_SKIP_DOWNLOAD true
ENV PUPPETEER_SKIP_CHROME_DOWNLOAD true
ENV PUPPETEER_SKIP_CHROME_HEADLESS_SHELL_DOWNLOAD true
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD true
RUN apt-get update && apt-get install gnupg wget -y && \
  wget --quiet --output-document=- https://dl-ssl.google.com/linux/linux_signing_key.pub | gpg --dearmor > /etc/apt/trusted.gpg.d/google-archive.gpg && \
  sh -c 'echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google.list' && \
  apt-get update && \
  apt-get install google-chrome-stable -y --no-install-recommends && \
  rm -rf /var/lib/apt/lists/*

ENV NODE_OPTIONS="--max-old-space-size=3072"

WORKDIR /app

COPY . /app

RUN npm i -g pnpm

RUN pnpm i --frozen-lockfile --prod false

ENTRYPOINT cd /app && pnpm start


### Describe the Bug

After version 10.6.1, it gives me error in my CI _pipeline_

Step 12/17 : RUN pnpm i --frozen-lockfile --prod false
 ---> Running in 15f858b90749
Scope: all 12 workspace projects
 ERR_PNPM_OPTIONAL_DEPS_REQUIRE_PROD_DEPS  Optional dependencies cannot be installed without production dependencies
The command '/bin/sh -c pnpm i --frozen-lockfile --prod false' returned a non-zero code: 1

### Expected Behavior

Normally it would have installed all packages properly

### Which Node.js version are you using?

22

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [x] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

I got the same issue. It worked just fine a few days ago when using pnpm 10.5.2

Here some logs from my Digital Ocean App Service deployment:

```bash
[2025-03-07 11:52:06] ╭──────────── buildpack detection ───────────╼
[2025-03-07 11:52:06] │  › using Ubuntu 22.04 stack
[2025-03-07 11:52:07] │ Detected the following buildpacks suitable to build your app:
[2025-03-07 11:52:07] │ 
[2025-03-07 11:52:07] │    digitalocean/nodejs-appdetect  v0.0.5    
[2025-03-07 11:52:07] │    heroku/nodejs                  v0.282.4  (Node.js)
[2025-03-07 11:52:07] │    digitalocean/procfile          v0.0.5    (Procfile)
[2025-03-07 11:52:07] │    digitalocean/custom            v0.1.3    (Custom Build Command)
[2025-03-07 11:52:07] ╰─────────────────────────────────────────────╼

... 

[2025-03-07 11:52:17] │ -----> Installing dependencies
[2025-03-07 11:52:17] │        Running 'pnpm install' with pnpm-lock.yaml
[2025-03-07 11:52:17] │        Scope: all 3 workspace projects
[2025-03-07 11:52:17] │         ERR_PNPM_OPTIONAL_DEPS_REQUIRE_PROD_DEPS  Optional dependencies cannot be installed without production dependencies
```

This is causing errors with the Heroku nodejs buildpack.

```
-----> Installing dependencies
       Running 'pnpm install' with pnpm-lock.yaml
       Scope: all 11 workspace projects
        ERR_PNPM_OPTIONAL_DEPS_REQUIRE_PROD_DEPS  Optional dependencies cannot be installed without production dependencies
-----> Build failed
```

--- a/.changeset/fresh-rooms-relate.md
+++ b/.changeset/fresh-rooms-relate.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/config": patch
+"pnpm": patch
+---
+
+`pnpm install --prod=false` should not crash, when executed in a project with a `pnpm-workspace.yaml` file [#9233](https://github.com/pnpm/pnpm/issues/9233). This fixes regression introduced via [#9211](https://github.com/pnpm/pnpm/pull/9211).

--- a/config/config/src/index.ts
+++ b/config/config/src/index.ts
@@ -346,18 +346,6 @@ export async function getConfig (opts: {
 
   pnpmConfig.packageManager = packageManager
 
-  if (pnpmConfig.only === 'prod' || pnpmConfig.only === 'production' || !pnpmConfig.only && pnpmConfig.production) {
-    pnpmConfig.production = true
-    pnpmConfig.dev = false
-  } else if (pnpmConfig.only === 'dev' || pnpmConfig.only === 'development' || pnpmConfig.dev) {
-    pnpmConfig.production = false
-    pnpmConfig.dev = true
-    pnpmConfig.optional = false
-  } else {
-    pnpmConfig.production = true
-    pnpmConfig.dev = true
-  }
-
   if (typeof pnpmConfig.filter === 'string') {
     pnpmConfig.filter = (pnpmConfig.filter as string).split(' ')
   }
@@ -510,6 +498,18 @@ export async function getConfig (opts: {
   }
 
   pnpmConfig.failedToLoadBuiltInConfig = failedToLoadBuiltInConfig
+
+  if (pnpmConfig.only === 'prod' || pnpmConfig.only === 'production' || !pnpmConfig.only && pnpmConfig.production) {
+    pnpmConfig.production = true
+    pnpmConfig.dev = false
+  } else if (pnpmConfig.only === 'dev' || pnpmConfig.only === 'development' || pnpmConfig.dev) {
+    pnpmConfig.production = false
+    pnpmConfig.dev = true
+    pnpmConfig.optional = false
+  } else {
+    pnpmConfig.production = true
+    pnpmConfig.dev = true
+  }
 
   return { config: pnpmConfig, warnings }
 }

